### PR TITLE
Upgrade dbt and sqlfluff, and adjust dbt CI

### PR DESCRIPTION
**Pull request by atvaccaro**
# Description

Upgrades dbt and sqlfluff; should solve the current CI version conflict issue that we've seen and brings in some recent updates around Python models in GCP.

Also gets us on sqlfluff v2 which has a lot of improvements.

Merge after https://github.com/cal-itp/data-infra/pull/2421

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] agencies.yml

## How has this been tested?
Tested dbt models; CI should pass.

### CODE REVIEW

**Review by lauriemerrell — APPROVED** [review]
Looks ok to me, would prefer if @SorenSpicknall also reviews. Also a bit confused about the merge deps, if it needs to merge after #2421 should it just be merging to main with a `do not merge` label?

**Comment by atvaccaro** [comment]
> Looks ok to me, would prefer if @SorenSpicknall also reviews. Also a bit confused about the merge deps, if it needs to merge after #2421 should it just be merging to main with a `do not merge` label?

This was originally one big PR so I wanted to break the commits up but keep the order in case of merge conflicts.

File: .pre-commit-config.yaml
Comment (SorenSpicknall) [inline .pre-commit-config.yaml:38]: 1.3.0 was previously set here because of issues with [breaking changes to exception names beginning in 1.4.0](https://github.com/dbt-labs/dbt-bigquery/pull/460/files) - is that handled?

Comment (atvaccaro) [reply .pre-commit-config.yaml:38]: Yes it was fixed in https://github.com/sqlfluff/sqlfluff/pull/4317

**Pull request merged**
